sat/cli/bootprep/input/image.py

```diff
@@ -479,3 +493,124 @@ def completed_successfully(self):
+        # is_recipe is a required property
+        return self.image_data['product']['is_recipe']
+
+    @property
+    def product_name(self):
+        """str: the product name"""
+        # name is a required property
+        return self.image_data['product']['name']
+
+    @property
+    @jinja_rendered
+    def product_version(self):
+        """str: the product version"""
+        # version is not required and should default to the latest version
+        return self.image_data['product'].get('version')
+
+    @property
+    def base_description(self):
+        """str: a human-readable description of the base we are starting from"""
+        return (
+            f'{self.base_resource_type} provided by '
+            f'{f"version {self.product_version}" if self.product_version else "latest version"} '
+            f'of product {self.product_name}'
+        )
+
+    @cached_property
+    def installed_product(self):
+        """cray_product_catalog.query.InstalledProductVersion: the matching product version"""
+        try:
+            return self.product_catalog.get_product(self.product_name, version=self.product_version)
+        except ProductCatalogError as err:
+            raise ImageCreateError(f'Failed to find {self.base_description}: {err}')
+
+    @cached_property
+    def ims_base(self):
+        """dict: the data for the base IMS recipe or image to build and/or customize"""
+        if self.base_is_recipe:
+            ims_resources = self.installed_product.recipes
+        else:
+            ims_resources = self.installed_product.images
+
+        if not ims_resources:
+            raise ImageCreateError(f'There is no {self.base_description}.')
+        elif len(ims_resources) > 1:
+            raise ImageCreateError(f'There exists more than one {self.base_description}.')
+        ims_resource = ims_resources[0]
+        ims_resource_desc = ', '.join(f'{key}={value}' for key, value in ims_resource.items())
+
+        try:
+            matching_base_resources = self.ims_client.get_matching_resources(
+                self.base_resource_type, resource_id=ims_resource['id']
+            )
+        except KeyError:
+            raise ImageCreateError(
+                f'{self.base_description} with {ims_resource_desc} does not '
+                f'have an id in product catalog data.'
+            )
+        except APIError as err:
+            raise ImageCreateError(err)
+
+        if len(matching_base_resources) == 0:
+            raise ImageCreateError(f'Found no matches in IMS for {self.base_description} '
+                                   f'with {ims_resource_desc}')
+        elif len(matching_base_resources) > 1:
+            # This really shouldn't happen given that we are querying by unique IMS id
+            raise ImageCreateError(f'Found {len(matching_base_resources)} matches for '
+                                   f'{self.base_description} with {ims_resource_desc}')
```

Comment:
nit: irrespective of my comment about structural pattern matching above, it might be good for now to have a consistent style for these sorts of conditionals checking if the list has exactly 1 element. Specifically, there are instances of both checking `len(some_list) == 0` and `not some_list` in this code. FWIW I typically prefer the latter style, but since we're specifically checking the length here I see the rationale for using the former. I don't feel strongly, but I do think we should be consistent regardless.

Reply:
I don't see a specific reason why I need to use `len(matching_base_resources) == 0` here vs. `not matching_base_resources`, so I will use the latter.